src/SDK/Common/InstrumentationScope/Configurator.php

```diff
@@ -0,0 +1,41 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OpenTelemetry\SDK\Common\InstrumentationScope;
+
+use OpenTelemetry\SDK\Common\Instrumentation\InstrumentationScopeInterface;
+
+class Configurator implements ScopeConfigurator
+{
+    /**
+     * @param list<Condition> $conditions
+     */
+    public function __construct(private readonly array $conditions = [])
+    {
+    }
+
+    /**
+     * @internal
+     */
+    public function getConfig(InstrumentationScopeInterface $scope): Config
+    {
+        foreach ($this->conditions as $condition) {
+            if ($condition->matches($scope)) {
+                return new Config($condition->state());
```

Comment:
I think that we do, or will, need a config class per signal. I was just trying to avoid copy-pasting the same code when they currently have the same functionality.
I don't mind addressing that sooner than later - either I do just turn the one general class into 3 signal-specific ones, or I'm happy to incorporate your alternative if you've got some more code to point me at?

Reply:
I was using the following `Configurator` implementation, with `TracerProvider`/`MeterProvider`/`LoggerProvider` accepting a `Closure(InstrumentationScope): TConfig` as configurator ("[Tracer/Meter/Logger]Configurator is modeled as a function to maximize flexibility.").

<details>
<summary>Configurator implementation</summary>

I'm not too happy with the state of this implementation as it does not support removal of callbacks yet.
Might want to add a predicate parameter to `::with()` that allows flexible conditions.

```php
/**
 * @template T
 */
final class Configurator {

    /** @var Closure(InstrumentationScope): T */
    private readonly Closure $factory;
    /** @var WeakMap<InstrumentationScope, T> */
    private WeakMap $configs;
    /** @var list<ConfiguratorClosure> */
    private array $configurators = [];

    /**
     * @param Closure(InstrumentationScope): T $factory
     */
    public function __construct(Closure $factory) {
        $this->configs = new WeakMap();
        $this->factory = $factory;
    }

    /**
     * @param Closure(T, InstrumentationScope): void $closure
     */
    public function with(Closure $closure, ?string $name, ?string $version = null, ?string $schemaUrl = null): self {
        $this->configurators[] = $configurator = new ConfiguratorClosure($closure, self::namePattern($name), $version, $schemaUrl);

        foreach ($this->configs as $instrumentationScope => $config) {
            if ($configurator->matches($instrumentationScope)) {
                ($configurator->closure)($config, $instrumentationScope);
            }
        }

        return $this;
    }

    /**
     * @return T
     */
    public function resolve(InstrumentationScope $instrumentationScope): object {
        if ($config = $this->configs[$instrumentationScope] ?? null) {
            return $config;
        }

        $config = ($this->factory)($instrumentationScope);
        foreach ($this->configurators as $configurator) {
            if ($configurator->matches($instrumentationScope)) {
                ($configurator->closure)($config, $instrumentationScope);
            }
        }

        return $this->configs[$instrumentationScope] ??= $config;
    }

    private static function namePattern(?string $name): ?string {
        return $name !== null
            ? sprintf('/^%s$/', strtr(preg_quote($name, '/'), ['\\?' => '.', '\\*' => '.*']))
            : null;
    }
}

/**
 * @internal
 */
final class ConfiguratorClosure {

    public function __construct(
        public readonly Closure $closure,
        private readonly ?string $name,
        private readonly ?string $version,
        private readonly ?string $schemaUrl,
    ) {}

    public function matches(InstrumentationScope $instrumentationScope): bool {
        return ($this->name === null || preg_match($this->name, $instrumentationScope->name))
            && ($this->version === null || $this->version === $instrumentationScope->version)
            && ($this->schemaUrl === null || $this->schemaUrl === $instrumentationScope->schemaUrl);
    }
}
```
```php
$meterProviderBuilder->setMeterConfigurator((new Configurator(static fn() => new MeterConfig()))
    ->with(static fn(MeterConfig $config) => $config->setDisabled(true), name: '*')
    ->with(static fn(MeterConfig $config) => $config->setDisabled(false), name: 'test')
    ->resolve(...)):
```
</details>

Reply:
Updated the Configurator implementation to use the above. I had to suppress a couple of phan/phpstan complaints about templates, but it works. I added a couple of convenience methods to create a default Configurator for each signal (`Configurator::meter()` etc).
I think it's also worth noting/documenting that if there are multiple matching rules in a configurator, then the last one wins.